# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxFcPlatformFontList.cpp
# Commit: 2378d74d6ae5
# Full Hash: 2378d74d6ae5e0411d5378e0f72e5552e4be7cde
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2022-03-10 17:23:56
# Description:
#   Bug 1758286 - Revert to separate Lookup and InsertOrUpdate calls when working with mFcSubstituteCache. r=lsalzman
#   
#   This shouldn't change behavior at all, afaict, but by reverting to the original
#   code structure I'm hoping it may avoid the release-assert that has showed up
#   in a few crash reports.
# ==============================================================================

diff -r 50bd11c68536 -r 2378d74d6ae5 gfx/thebes/gfxFcPlatformFontList.cpp
--- a/gfx/thebes/gfxFcPlatformFontList.cpp	Thu Mar 10 00:58:07 2022 +0000
+++ b/gfx/thebes/gfxFcPlatformFontList.cpp	Thu Mar 10 01:06:56 2022 +0000
@@ -2113,63 +2113,73 @@
   ToLowerCase(cacheKey);
   cacheKey.Append(':');
   cacheKey.Append(familyName);
-  const auto& cached = mFcSubstituteCache.LookupOrInsertWith(cacheKey, [&] {
-    // It wasn't in the cache, so we need to ask fontconfig...
-    const FcChar8* kSentinelName = ToFcChar8Ptr("-moz-sentinel");
-    const FcChar8* terminator = nullptr;
-    RefPtr<FcPattern> sentinelSubst = dont_AddRef(FcPatternCreate());
-    FcPatternAddString(sentinelSubst, FC_FAMILY, kSentinelName);
-    if (!lang.IsEmpty()) {
-      FcPatternAddString(sentinelSubst, FC_LANG, ToFcChar8Ptr(lang.get()));
+  if (const auto& cached = mFcSubstituteCache.Lookup(cacheKey)) {
+    if (cached->IsEmpty()) {
+      return false;
     }
-    FcConfigSubstitute(nullptr, sentinelSubst, FcMatchPattern);
+    aOutput->AppendElements(*cached);
+    return true;
+  }
+
+  // It wasn't in the cache, so we need to ask fontconfig...
+  const FcChar8* kSentinelName = ToFcChar8Ptr("-moz-sentinel");
+  const FcChar8* terminator = nullptr;
+  RefPtr<FcPattern> sentinelSubst = dont_AddRef(FcPatternCreate());
+  FcPatternAddString(sentinelSubst, FC_FAMILY, kSentinelName);
+  if (!lang.IsEmpty()) {
+    FcPatternAddString(sentinelSubst, FC_LANG, ToFcChar8Ptr(lang.get()));
+  }
+  FcConfigSubstitute(nullptr, sentinelSubst, FcMatchPattern);
 
-    // If the sentinel name is still present, we'll use that as the terminator
-    // for the family names we collect; this means that if fontconfig prepends
-    // additional family names (e.g. an emoji font, or lang-specific preferred
-    // font) to all patterns, it won't simply mask all actual requested names.
-    // If the sentinel has been deleted/replaced altogether, then we'll take
-    // the first substitute name as the new terminator.
-    FcChar8* substName;
-    for (int i = 0; FcPatternGetString(sentinelSubst, FC_FAMILY, i,
-                                       &substName) == FcResultMatch;
-         i++) {
-      if (FcStrCmp(substName, kSentinelName) == 0) {
-        terminator = kSentinelName;
-        break;
-      } else if (!terminator) {
-        terminator = substName;
-      }
+  // If the sentinel name is still present, we'll use that as the terminator
+  // for the family names we collect; this means that if fontconfig prepends
+  // additional family names (e.g. an emoji font, or lang-specific preferred
+  // font) to all patterns, it won't simply mask all actual requested names.
+  // If the sentinel has been deleted/replaced altogether, then we'll take
+  // the first substitute name as the new terminator.
+  FcChar8* substName;
+  for (int i = 0; FcPatternGetString(sentinelSubst, FC_FAMILY, i, &substName) ==
+                  FcResultMatch;
+       i++) {
+    if (FcStrCmp(substName, kSentinelName) == 0) {
+      terminator = kSentinelName;
+      break;
     }
+    if (!terminator) {
+      terminator = substName;
+    }
+  }
 
-    // substitutions for font, -moz-sentinel pattern
-    RefPtr<FcPattern> fontWithSentinel = dont_AddRef(FcPatternCreate());
-    FcPatternAddString(fontWithSentinel, FC_FAMILY,
-                       ToFcChar8Ptr(familyName.get()));
-    FcPatternAddString(fontWithSentinel, FC_FAMILY, kSentinelName);
-    if (!lang.IsEmpty()) {
-      FcPatternAddString(sentinelSubst, FC_LANG, ToFcChar8Ptr(lang.get()));
-    }
-    FcConfigSubstitute(nullptr, fontWithSentinel, FcMatchPattern);
+  // substitutions for font, -moz-sentinel pattern
+  RefPtr<FcPattern> fontWithSentinel = dont_AddRef(FcPatternCreate());
+  FcPatternAddString(fontWithSentinel, FC_FAMILY,
+                     ToFcChar8Ptr(familyName.get()));
+  FcPatternAddString(fontWithSentinel, FC_FAMILY, kSentinelName);
+  if (!lang.IsEmpty()) {
+    FcPatternAddString(sentinelSubst, FC_LANG, ToFcChar8Ptr(lang.get()));
+  }
+  FcConfigSubstitute(nullptr, fontWithSentinel, FcMatchPattern);
 
-    // Add all font family matches until reaching the terminator.
-    AutoTArray<FamilyAndGeneric, 10> cachedFamilies;
-    for (int i = 0; FcPatternGetString(fontWithSentinel, FC_FAMILY, i,
-                                       &substName) == FcResultMatch;
-         i++) {
-      if (terminator && FcStrCmp(substName, terminator) == 0) {
-        break;
-      }
-      gfxPlatformFontList::FindAndAddFamilies(
-          aPresContext, aGeneric, nsDependentCString(ToCharPtr(substName)),
-          &cachedFamilies, aFlags, aStyle, aLanguage);
+  // Add all font family matches until reaching the terminator.
+  AutoTArray<FamilyAndGeneric, 10> cachedFamilies;
+  for (int i = 0; FcPatternGetString(fontWithSentinel, FC_FAMILY, i,
+                                     &substName) == FcResultMatch;
+       i++) {
+    if (terminator && FcStrCmp(substName, terminator) == 0) {
+      break;
     }
-    return cachedFamilies;
-  });
-  if (cached.IsEmpty()) {
+    gfxPlatformFontList::FindAndAddFamilies(
+        aPresContext, aGeneric, nsDependentCString(ToCharPtr(substName)),
+        &cachedFamilies, aFlags, aStyle, aLanguage);
+  }
+
+  const auto& insertedCachedFamilies =
+      mFcSubstituteCache.InsertOrUpdate(cacheKey, std::move(cachedFamilies));
+
+  if (insertedCachedFamilies.IsEmpty()) {
     return false;
   }
-  aOutput->AppendElements(cached);
+  aOutput->AppendElements(insertedCachedFamilies);
   return true;
 }
 
